### PR TITLE
QS util cleanup

### PR DESCRIPTION
**Pull request by keithjgrant**
##### SUMMARY
Refactor much of the util/qs.js utilities file and increase test coverage to 100%.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
 - UI

##### ADDITIONAL INFORMATION
* Reordered/reorganized the qs.js file so functions are closer to where they are called for better readability -- Note this does make the diff a bit more confusing to look at, sorry :/
* Rewrote parts that were confusing or had old/unused logic (removed about 150 lines from the file total)
* Increased test coverage to 100%
* Replaced `addParams` function with `mergeParams` and `replaceParams` to make the API more explicit (removes "magic" where some params were always merged, others were always replaced)

**The big thing I'm going for here is READABILITY and making these utils intuitive to use, so please flag anything that's not fairly easy to understand quickly!**

### CODE REVIEW

File: awx/ui_next/src/util/qs.js [hunk 1]
```diff
@@ -1,54 +1,59 @@
...
+        : decodeURIComponent(nsKey);
+      const value = parseValue(config, key, rawValue);
+      params[key] = mergeParam(params[key], value);
+    });
+  return params;
+}
+export { stringToObject as _stringToObject };
 
 /**
  * helper function to check the namespace of a param is what you expec
```
Comment (AlexSCorey) [inline awx/ui_next/src/util/qs.js]: missing a `t` here at the end of expect.

File: awx/ui_next/src/util/qs.test.js [hunk 2]
```diff
@@ -540,12 +511,237 @@ describe('qs (qs.js)', () => {
...
+      };
+      const newParams = {
+        foo: 'three',
+      };
+      expect(mergeParams(oldParams, newParams)).toEqual({
+        foo: ['one', 'two', 'three'],
+      });
+    });
+
+    it('should append array to existing value', () => {
```
Comment (AlexSCorey) [inline awx/ui_next/src/util/qs.test.js:650]: Is this test needed?  It seems to be testing the same thing as the test above.

File: awx/ui_next/src/util/qs.test.js [hunk 3]
```diff
@@ -540,12 +511,237 @@ describe('qs (qs.js)', () => {
...
+      };
+      const newParams = {
+        foo: ['two', 'three'],
+      };
+      expect(mergeParams(oldParams, newParams)).toEqual({
+        foo: ['one', 'two', 'three'],
+      });
+    });
+
+    it('should merge two arrays', () => {
```
Comment (AlexSCorey) [inline awx/ui_next/src/util/qs.test.js:662]: Same here. Is it needed?

**Review by AlexSCorey — CHANGES_REQUESTED** [review]
There is a typo in a comment, not a huge deal but thought I'd point it out. I also had just a few questions.

File: awx/ui_next/src/util/qs.test.js (hunk 2 above)
Comment (keithjgrant) [reply awx/ui_next/src/util/qs.test.js:650]: they are three slightly different cases: val + val => array, array + val => array, val + array => array, each of which exercise different paths of the code

**Review by AlexSCorey — APPROVED** [review]

**Review by jlmitch5 — APPROVED** [review]
I really like how qs reads now.  We worked out the add -> merge/replace earlier, and I like that a lot better too.  Awesome stuff!

**Pull request merged**